README.md
# oskr-email

A companion service that handles sending for delivery for [oskr-events](https://github.com/orgsync/oskr-events).

## Credits

oskr-email is brought to you by [@hadronzoo][1] and [@larslevie][2] from
[@OrgSync][3].

[1]: http://github.com/hadronzoo
[2]: http://github.com/larslevie
[3]: http://github.com/orgsync

## License

Copyright 2016 OrgSync.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-09-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-09-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/oskr-email/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/oskr-email/compare/0.1.0...0.1.1


project.clj
;
; Copyright 2016 OrgSync.
;
; Licensed under the Apache License, Version 2.0 (the "License")
; you may not use this file except in compliance with the License.
; You may obtain a copy of the License at
;
;   http://www.apache.org/licenses/LICENSE-2.0
;
; Unless required by applicable law or agreed to in writing, software
; distributed under the License is distributed on an "AS IS" BASIS,
; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
; See the License for the specific language governing permissions and
; limitations under the License.
;

(defproject oskr-email "0.1.0"
  :description "A companion to oskr-events that delivers messages via email"
  :url "http://github.com/orgsync/oskr-email"
  :license {:name "Apache 2.0"
            :url  "http://www.apache.org/licenses/LICENSE.md-2.0"}
  :dependencies [[byte-streams "0.2.2"]
                 [cheshire "5.6.3"]
                 [com.draines/postal "2.0.1"]
                 [com.stuartsierra/component "0.3.1"]
                 [environ "1.1.0"]
                 [log4j/log4j "1.2.17"]
                 [manifold "0.1.6-alpha1"]
                 [net.cgrand/xforms "0.4.0"]
                 [org.apache.kafka/kafka-clients "0.9.0.1"]
                 [org.clojure/clojure "1.8.0"]
                 [org.clojure/tools.logging "0.3.1"]
                 [org.slf4j/slf4j-log4j12 "1.7.21"]
                 [org.slf4j/slf4j-api "1.7.21"]]
  :profiles {:repl    {:jvm-opts ["-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"]}
             :uberjar {:jvm-opts     ["-Dclojure.compiler.direct-linking=true"]
                       :aot          :all
                       :uberjar-name "oskr-email.jar"}}
  :main oskr-email.core)


src/oskr_email/core.clj
;
; Copyright 2016 OrgSync.
;
; Licensed under the Apache License, Version 2.0 (the "License")
; you may not use this file except in compliance with the License.
; You may obtain a copy of the License at
;
;   http://www.apache.org/licenses/LICENSE-2.0
;
; Unless required by applicable law or agreed to in writing, software
; distributed under the License is distributed on an "AS IS" BASIS,
; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
; See the License for the specific language governing permissions and
; limitations under the License.
;

(ns oskr-email.core
  (:require [com.stuartsierra.component :as component]
            [oskr-email.processor :as processor]
            [oskr-email.consumer :as consumer]
            [manifold.deferred :as d]
            [environ.core :refer [env]]
            [clojure.tools.logging :refer [error]])
  (:gen-class))

(defn create-system []
  (let [kafka-bootstrap (env "KAFKA_BOOTSTRAP" "kafka.service.consul:9092")]
    (-> (component/system-map
          :processor (processor/new-processor)
          :consumer (consumer/new-consumer
                      kafka-bootstrap
                      (env "KAFKA_GROUP_ID" "oskr-email")
                      (env "KAFKA_SPEC_TOPIC" "Communications.Deliveries.Email")))
        (component/system-using
          {:processor [:consumer]}))))

(def default-system
  (create-system))

(defn -main [& _]
  (let [system (component/start default-system)]
    (Thread/setDefaultUncaughtExceptionHandler
      (reify Thread$UncaughtExceptionHandler
        (uncaughtException [_ thread ex]
          (error "Uncaught exception on" (.getName thread) "-"
                 (.getMessage ex)))))

    (.addShutdownHook (Runtime/getRuntime)
                      (Thread. ^Runnable #(component/stop system)))

    @(d/deferred)))

(comment
  (do
    (def s (create-system))
    (alter-var-root #'s component/start)))

(comment
  (alter-var-root #'s component/stop))


src/oskr_email/processor.clj
;
; Copyright 2016 OrgSync.
;
; Licensed under the Apache License, Version 2.0 (the "License")
; you may not use this file except in compliance with the License.
; You may obtain a copy of the License at
;
;   http://www.apache.org/licenses/LICENSE-2.0
;
; Unless required by applicable law or agreed to in writing, software
; distributed under the License is distributed on an "AS IS" BASIS,
; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
; See the License for the specific language governing permissions and
; limitations under the License.
;

(ns oskr-email.processor
  (:require [clojure.tools.logging :refer [info debug error warn]]
            [com.stuartsierra.component :as component]
            [environ.core :refer [env]]
            [manifold.stream :as s]
            [manifold.deferred :as d]
            [net.cgrand.xforms :as x]
            [oskr-email.protocols :as p]
            [postal.smtp :refer [smtp-send]]))

(defn ack-deliveries [consumer message-batch]
  (let [acks (into {} (comp
                        (map meta)
                        (x/by-key :partition (comp
                                               (map :offset)
                                               (x/reduce max 0))))
                   message-batch)]
    (doseq [[partition offset] acks]
      (p/commit! consumer partition offset))))

(defn deliver! [message-batch]
  (info "Starting delivery")

  (let [host (env "SMTP_HOST" "mailcatcher.service.consul")
        port (env "SMTP_PORT" 1025)]
    (apply smtp-send (list* {:host host :port port} message-batch))))

(defn make-delivery-handler [consumer]
  (fn [message-batch]
    (info "handling delivery")
    (d/chain' (deliver! message-batch)
      (fn [_] (ack-deliveries consumer message-batch)))))

(defn process [message-stream consumer]
  (let [delivery-handler (make-delivery-handler consumer)]
    (info "Starting process loop")
    (-> (d/loop []
          (d/let-flow [message-batch (s/take! message-stream)]
            (info "message-batch" message-batch)
            (if message-batch
              (d/chain' (delivery-handler message-batch)
                (fn [_] (d/recur))))))
        (d/catch' #(error %)))))

(defrecord Processor [consumer batched-stream finished?]
  component/Lifecycle
  (start [processor]
    (info "Starting processor")
    (let [message-stream (:message-stream consumer)
          batched-stream (s/batch 1000 200 message-stream)
          finished? (process batched-stream consumer)]
      (assoc processor :message-stream message-stream
                       :batched-stream batched-stream
                       :finished? finished?)))

  (stop [processor]
    (info "Stopping processor")
    (s/close! batched-stream)
    @finished?
    (assoc processor :batched-stream nil
                     :finished? nil)))

(defn new-processor []
  (Processor. nil nil nil))

#_(comment
    (let [offset-atom (atom 0)
          messages (as-> "resources/deliveries.json" $
                         (java.io.File. $)
                         (slurp $)
                         (json/parse-string $ true)
                         (map delivery->message $)
                         (map (fn [message]
                                (with-meta message {:topic     "Deliveries"
                                                    :partition (rand-int 5)
                                                    :offset    (swap! offset-atom inc)}))
                              $))]
      (def msgs messages)

      (do
        (deliver! messages))))


src/oskr_email/consumer.clj
;
; Copyright 2016 OrgSync.
;
; Licensed under the Apache License, Version 2.0 (the "License")
; you may not use this file except in compliance with the License.
; You may obtain a copy of the License at
;
;   http://www.apache.org/licenses/LICENSE-2.0
;
; Unless required by applicable law or agreed to in writing, software
; distributed under the License is distributed on an "AS IS" BASIS,
; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
; See the License for the specific language governing permissions and
; limitations under the License.
;

(ns oskr-email.consumer
  (:require [byte-streams :as byte]
            [clojure.tools.logging :refer [info debug error warn]]
            [clojure.walk :refer [stringify-keys]]
            [com.stuartsierra.component :as component]
            [environ.core :refer [env]]
            [manifold.stream :as s]
            [oskr-email.protocols :as p]
            [cheshire.core :as json])
  (:import [org.apache.kafka.common.serialization ByteArrayDeserializer StringDeserializer]
           [org.apache.kafka.clients.consumer KafkaConsumer ConsumerRecord OffsetAndMetadata OffsetCommitCallback]
           [java.util Map]
           [java.io Reader]
           [org.apache.kafka.common TopicPartition]))

(defn record->message [^ConsumerRecord record]
  (let [content (-> (.value record)
                    (byte/convert Reader)
                    (json/parse-stream true)
                    :content)
        message {:to         (:to content)
                 :from       (:from content)
                 :subject    (:subject content)
                 :body       (:content content)
                 :message-id (constantly (:deliverableId content))
                 :user-agent "Oskr Email"}]
    (with-meta message {:topic     (.topic record)
                        :partition (.partition record)
                        :offset    (.offset record)})))

(defn poll-fn [kafka-consumer message-stream]
  (info "defining poller")
  (fn []
    (when-not (s/closed? message-stream)
      (warn "polling")
      (->> (try (locking kafka-consumer
                  (.poll kafka-consumer 1000))
                (catch Exception e
                  (warn "polling failed" (.getMessage e))))
           (.iterator)
           (iterator-seq)
           (s/put-all! message-stream)
           deref)
      (recur))))

(defn spawn-poller [kafka-consumer message-stream]
  (info "spawning poller")
  (-> (Thread. ^Runnable (poll-fn kafka-consumer message-stream)
               "kafka-consumer")
      (.start)))

(defrecord Consumer [config topic message-stream kafka-consumer]
  component/Lifecycle
  (start [consumer]
    (info "Starting Consumer")
    (let [message-stream (s/stream* {:xform (map record->message)})
          kafka-consumer (try (KafkaConsumer. ^Map config)
                              (catch Exception e (warn e)))]
      (spawn-poller kafka-consumer message-stream)
      (locking kafka-consumer
        (.assignment kafka-consumer)
        (.subscribe kafka-consumer [topic]))
      (assoc consumer :kafka-consumer kafka-consumer
                      :message-stream message-stream)))

  (stop [consumer]
    (try
      (s/close! message-stream)
      (locking kafka-consumer
        (.close kafka-consumer))
      (assoc consumer :kafka-consumer nil
                      :message-stream nil)
      (catch Exception e (error e)
                         (throw e))))

  p/Commitable
  (commit! [_ partition offset]
    (info "acking to consumer")
    (let [topic (env "KAFKA_TOPIC" "Communications.Deliveries.Email")
          topic-partition (TopicPartition. topic partition)
          offset-and-metadata (OffsetAndMetadata. (inc offset))
          offsets {topic-partition offset-and-metadata}
          callback (reify OffsetCommitCallback
                     (onComplete [_ _ e]
                       (when e
                         (warn "Unable to commit"
                               [topic partition offset]
                               (.getMessage e)))))]

      (locking kafka-consumer
        (.commitAsync kafka-consumer offsets callback)))))

(defn new-consumer [kafka-bootstrap group-id topic]
  (let [config (-> {:bootstrap.servers  kafka-bootstrap
                    :key.deserializer   StringDeserializer
                    :value.deserializer ByteArrayDeserializer
                    :group.id           group-id
                    :enable.auto.commit false
                    :auto.offset.reset  "earliest"}
                   stringify-keys)]
    (Consumer. config topic nil nil)))

(comment
  (def consumer (new-consumer "kafka.service.consul:9092"
                              "oskr-email"
                              "Communications.Deliveries.Email")))

(comment
  (alter-var-root #'consumer component/start))

(comment
  (alter-var-root #'consumer component/stop))

#_(comment
    (let [consumer (new-consumer "kafka.service.consul:9092"
                                 "oskr-email"
                                 "Communications.Deliveries.Email")
          offset-atom (atom 0)
          messages (as-> "resources/deliveries.json" $
                         (java.io.File. $)
                         (slurp $)
                         (json/parse-string $ true)
                         (map #(ConsumerRecord.
                                "Communications.Deliveries.Email"
                                (rand-int 5)
                                (swap! offset-atom inc)
                                "iosdyfu938" %) $)
                         (map #(record->message %) $))]
      (clojure.pprint/pprint messages)))
